8254354: Add a withInvokeExactBehavior() VarHandle combinator

JornVernee opened the pull request:
Hi,

This patch adds an asExact() combinator to VarHandle, that will return a new VarHandle that performs exact type checks, similar to MethodHandle::invokeExact, to help developers catch inexact VarHandle usage, which can lead to performance degradation.

This is implemented using a boolean flag in VarForm. If the flag is set, the exact type of the invocation is checked against the exact type in the VarForm. If there is a mismatch, a WrongMethodTypeException is thrown.

Other than that, there is also an asGeneric() combinator added that does the inverse operation (thanks to Rémi for the suggestion). I've also added The `@Hidden` annotation to the VarHandleGuards methods, as well as a type-checking helper method called from the generic invocation lambda form, so that the stack trace we get points at the location where the VarHandle is being used.

Thanks,
Jorn

CSR link: https://bugs.openjdk.java.net/browse/JDK-8255375

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8254354](https://bugs.openjdk.java.net/browse/JDK-8254354): Add a withInvokeExactBehavior() VarHandle combinator


### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to 3c707bc73b971a82bef162e4cc077d1ef0e1d1e9
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to 3097f0800403be9bee7a82b1ef94b888be7c676c


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/843/head:pull/843`
`$ git checkout pull/843`

JornVernee commented:
/label remove hotspot-compiler

forax commented on src/java.base/share/classes/java/lang/invoke/MemoryAccessVarHandleGenerator.java:
```diff
@@ -418,6 +447,10 @@ private static int getSlotsForType(Class<?> c) {
         return 1;
     }
 
+    private String internalName(Class<?> cls) {
```
should be static, no ?

JornVernee replied:
Yes, thanks.

forax commented:
Looks good to me.

PaulSandoz reviewed (commented):
This approach does not work for reference types, since they are erased to `Object`, and then exact checking will be performed on the erased reference types.

For example try this:

```java
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;

public class Test {
    int x;

    public static void main(String[] args) throws Throwable {
        VarHandle x = MethodHandles.lookup().findVarHandle(Test.class, "x", int.class);
        VarHandle ex = x.asExact();
        Test t = new Test();
        ex.set(t, 1);
    }
}
```

Which results in:

```
Exception in thread "main" java.lang.invoke.WrongMethodTypeException: expected (Object,int)void but found (Test,int)void
	at Test.main(Test.java:11)
```

Exact type checking requires that match be performed on the VH access mode method type and the exact symbolic method type, something like:

```java
    final static Object guard_L_L(VarHandle handle, Object arg0, VarHandle.AccessDescriptor ad) throws Throwable {
        if (handle.vform.exact && handle.accessModeType(ad.type) != ad.symbolicMethodTypeExact) {
            throw new WrongMethodTypeException("expected " + handle.vform.methodType_table_exact[ad.type] + " but found "
                    + ad.symbolicMethodTypeExact);
        }
```

Then it's precisely the same as `MH.invokeExact`, rather than `MH.invoke`.

A `VarForm` is a resource shared across many instances of the same _kind_ of `VarHandle`, so cannot be used for exact matching, except in specific scenarios e.g. access on a primitive array.

JornVernee commented:
@PaulSandoz Thanks. I initially tested this with memory access VarHandles, which don't erase the receiver type. e.g.

```
MemoryLayout layout = MemoryLayout.ofSequence(10, JAVA_INT);
VarHandle vh = layout.varHandle(int.class, MemoryLayout.PathElement.sequenceElement());
vh = vh.asExact();
try (MemorySegment segment = MemorySegment.allocateNative(layout)) {
    for (int i = 0; i <10; i++) {
        vh.set(segment.baseAddress(), i, i);
    }
}
```

Will result in:
```
Exception in thread "main" java.lang.invoke.WrongMethodTypeException: expected (MemoryAddress,long,int)void but found (MemoryAddress,int,int)void
	at java.base/java.lang.invoke.VarHandleGuards.guard_LII_V(VarHandleGuards.java:915)
	at main.Main.main(Main.java:18)
```

Which led me to believe the approach would work for other reference types. But, I suppose the MethodTypes fed to memaccess VarForms are non-erased as an exception rather than a rule.

I'll update the patch and sharpen the tests to check that the actual expected type is correct (per the exception message).

JornVernee commented:
@PaulSandoz I've implemented your suggestion. FWIW, the VH::accessModeType method took and AccessMode value as an argument, and the AccessDescriptor only stored the ordinal, so I added an `@Stable` array of values to AccessMode to map from ordinal to enum value. But, maybe it makes more sense to just store the AccessMode in the AccessDescriptor directly? (instead of the ordinal). Not sure what the original motivation was for only storing the ordinal.

I've also sharpened the tests to check the exception message. Do you think the testing is sufficient? (Note that I did not add tests to the template files since only a select set of argument type conversions causes the WMTE we're looking for. So, that's why I created a new test file).

FWIW, there seems to have been a bug in the implementation of IndirectVarHandle::accessModeTypeUncached, where it was using the VarHandle's type as the receiver argument (unlike all the other impls). I've fixed this by passing `null` instead, and also removed a workaround for this problem in VarHandles::maybeAdapt.

JornVernee commented:
/csr needed

JornVernee commented:
I've updated the javadoc, and added two benchmarks that show the existing descrepency between an exact and a generic use of a VarHandle, as well as showing that an exact VarHandle is as fast as a generic VarHandle for an exact invocation. (1 benchmark for normal Java field access, and 1 benchmark for the foreign memory-access API).

PaulSandoz commented:
The direct use of the enum ordinal is because HotSpot accessing it from the enum value is (or was) not optimal in C2.
 
You can avoid the addition of the stable array by doing the following:

```java
    public final MethodType accessModeType(AccessMode accessMode) {
        return accessModeType(accessMode.at.ordinal());
    }

    @ForceInline
    final MethodType accessModeType(int accessModeOrdinal) {
        TypesAndInvokers tis = getTypesAndInvokers();
        MethodType mt = tis.methodType_table[accessModeOrdinal];
        if (mt == null) {
            mt = tis.methodType_table[accessModeOrdinal] =
                    accessModeTypeUncached(accessModeOrdinal);
        }
        return mt;
    }

    final MethodType accessModeTypeUncached(int accessModeOrdinal) {
        return accessModeTypeUncached(AccessMode.values()[accessModeOrdinal]);
    }
```
It's a little odd going back and forth between the ordinal and the enum type, but it's all on the slow uncached path, and it avoids some duplication of code.

I'll take a closer looks at the other areas and respond in another comment.

PaulSandoz commented:
We can clarify the new methods and tie them closer to method handle semantics. I suggest the names `asExactInvoker` and `asInvoker`, referencing `MethodHandles.exactInvoker` and `MethodHandles.invoker` respectively. (The term "generic" is really reserved for erased method types, and otherwise used internally as the generic invoker.)

The `VarHandle` documentation already specifies that:

```
 * <p>
 * Invocation of an access mode method behaves as if an invocation of
 * {@link MethodHandle#invoke}, where the receiving method handle accepts the
 * VarHandle instance as the leading argument.  More specifically, the
 * following, where {@code {access-mode}} corresponds to the access mode method
 * name:
 * <pre> {@code
 * VarHandle vh = ..
 * R r = (R) vh.{access-mode}(p1, p2, ..., pN);
 * }</pre>
 * behaves as if:
 * <pre> {@code
 * VarHandle vh = ..
 * VarHandle.AccessMode am = VarHandle.AccessMode.valueFromMethodName("{access-mode}");
 * MethodHandle mh = MethodHandles.varHandleExactInvoker(
 *                       am,
 *                       vh.accessModeType(am));
 *
 * R r = (R) mh.invoke(vh, p1, p2, ..., pN)
 * }</pre>
 * (modulo access mode methods do not declare throwing of {@code Throwable}).
...
```

 We will need to adjust this section, i can help, but first let us reach agreement on this approach.

forax commented:
Paul,
an invoker has the MethodHandle (resp. VarHandle) as first argument so it's not the same semantics.

JornVernee commented:
I've updated the implementation of accessModeType to work with the ordinal directly. Note that it was using the AccessType ordinal though, so I also had to change the parameter type of accessModeTypeUncached, and the respective implementations, to use AccessType directly (luckily this was possible, since the current implementations all just used the AccessType).

JornVernee commented:
I think `asInvoker` and `asExactInvoker` make sense if you think of a VarHandle as an invoker of MethodHandles (though this is more of an implementation detail). But, I feel like the name `asInvoker` isn't obvious enough for what it does. i.e. it's not that obvious from the name that this removes the exactness.

If `asGeneric` is problematic, maybe just `asNonExact` works?

JornVernee commented:
I've uploaded another revision that has the suggested javadoc changes, courtesy of Paul. We also decided to rename asExact() and asGeneric() to withInvokeExactBehaviour() and withInvokeBehaviour() (with links to the relevant javadoc sections).

PaulSandoz reviewed (approved):
Looks good. There is just one difference between the spec and implementation. The spec states: 
> If this VarHandle already has invoke{-exact} behaviour this VarHandle is returned.

I prefer this behaviour, but feel free to update the spec if you like e.g. If ... already has XXX then a new VH with exactly the same behaviour as this VH is returned.

I just realized i used "behaviour", the english spelling. We should use the US spelling, "behavior", which is more commonly used throughout the JDK documentation.

PaulSandoz commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -1541,6 +1590,37 @@ VarHandle asDirect() {
+     * If this VarHandle already has invoke-exact behavior this VarHandle is returned.
+     * @apiNote
+     * Invoke-exact behavior guarantees that that upon invocation of an access mode method
+     * the types an arity of the arguments must match the {@link #accessModeType(AccessMode) access mode type},
```
s/an arity/and arity/

ChrisHegarty commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -1541,6 +1590,37 @@ VarHandle asDirect() {
+     * <p>
+     * If this VarHandle already has invoke-exact behavior this VarHandle is returned.
+     * @apiNote
+     * Invoke-exact behavior guarantees that that upon invocation of an access mode method
```
minor typo "that that"

ChrisHegarty commented on src/java.base/share/classes/java/lang/invoke/VarHandle.java:
```diff
@@ -1541,6 +1590,37 @@ VarHandle asDirect() {
+     * Invoke-exact behavior guarantees that that upon invocation of an access mode method
+     * the types and arity of the arguments must match the {@link #accessModeType(AccessMode) access mode type},
+     * otherwise a {@link WrongMethodTypeException} is thrown.
+     *
```
While not strictly necessary, the pedantic part to me likes to see assertions like this in the spec:

"Invoking `hasInvokeExactBehavior` on the returned VarHandle is guaranteed to return true."  And conversely, for withInvokeBehavior.

JornVernee commented:
@ChrisHegarty Thanks for the comments. I've implemented them.

ChrisHegarty commented on test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java:
```diff
@@ -0,0 +1,416 @@
+                             SetStaticX staticSetter, GetStaticX staticGetter)
+            throws NoSuchFieldException, IllegalAccessException {
+        if (ro) throw new SkipException("Can not test setter with read only field");
+        VarHandle vh = MethodHandles.lookup().findVarHandle(Widget.class, fieldBaseName + "_RW", fieldType);
```
Quite trivially we can assertFalse(vh. hasInvokeExactBehavior())

ChrisHegarty commented on test/jdk/java/lang/invoke/VarHandles/VarHandleTestExact.java:
```diff
@@ -0,0 +1,416 @@
+            fail("Unexpected exception", wmte);
+        }
+
+        vh = vh.withInvokeExactBehavior();
```
Quite trivially we can assertTrue(vh. hasInvokeExactBehavior())

ChrisHegarty reviewed (approved):
LGTM.

The test could be sprinkled with a number of assertions related to the invocation behaviour, a.k.a `hasInvokeExactBehavior`; either the newly specified default behaviour assertFalse(vh. hasInvokeExactBehavior()), or post switching to exact: assertTrue(vh. hasInvokeExactBehavior())

JornVernee commented:
@ChrisHegarty Good suggestion! Added.

JornVernee commented:
/integrate
